components/base/wheeled/kinematics.go

```diff
@@ -0,0 +1,118 @@
+package wheeled
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/golang/geo/r3"
+	"github.com/pkg/errors"
+
+	"go.viam.com/rdk/components/base"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/pointcloud"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/services/slam"
+	"go.viam.com/rdk/spatialmath"
+)
+
+type kinematicWheeledBase struct {
+	*wheeledBase
+	slam  slam.Service
+	model referenceframe.Model
+}
+
+// WrapWithKinematics takes a wheeledBase component and adds a slam service to it
+// It also adds kinematic model so that it can be controlled.
+func WrapWithKinematics(ctx context.Context, base *wheeledBase, slam slam.Service) (base.KinematicBase, error) {
+	var err error
+	kwb := &kinematicWheeledBase{
+		wheeledBase: base,
+		slam:        slam,
+	}
+	kwb.model, err = kwb.buildModel(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return kwb, err
+}
+
+func (kwb *kinematicWheeledBase) ModelFrame() referenceframe.Model {
+	return kwb.model
+}
+
+func (kwb *kinematicWheeledBase) CurrentInputs(ctx context.Context) ([]referenceframe.Input, error) {
+	// TODO(RSDK-2311): complete the implementation
+	return []referenceframe.Input{}, errors.New("not implemented yet")
+}
+
+func (kwb *kinematicWheeledBase) GoToInputs(ctx context.Context, goal []referenceframe.Input) error {
+	// TODO(RSDK-2311): complete the implementation
+	return errors.New("not implemented yet")
+}
+
+// ModelFrame builds the kinematic model associated with the kinematicWheeledBase
+// Note that this model is not intended to be registered in the frame system.
+func (kwb *kinematicWheeledBase) buildModel(ctx context.Context) (referenceframe.Model, error) {
+	if kwb.collisionGeometry == nil {
+		return nil, errors.New("cannot create model for base with no collision geometry")
+	}
+
+	// get the limits of the SLAM map to set as the extents of the frame
+	data, err := slam.GetPointCloudMapFull(ctx, kwb.slam, "")
+	if err != nil {
+		return nil, err
+	}
+	dims, err := pointcloud.GetPCDMetaData(bytes.NewReader(data))
```

Comment:
Could this be:

```suggestion
	pointcloud, err := pointcloud.ReadPCD(bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	dims := pointcloud.MetaData()
```

& if so could we remove the changes to pointcloud_file.go?

I think it would be better if there were only one canonical way to get the dimensions of a pointcloud unless there is a massive performance or behavioral difference between `pointcloud.GetPCDMetaData` and my suggestion.

Reply:
https://github.com/viamrobotics/rdk/pull/2058#discussion_r1140474583

Reply:
I actually had this earlier but followed Tess's suggestion to change it for performance reasons